genwinsyscallproto/genwinsyscalldef.go

```diff
@@ -0,0 +1,311 @@
+	}
+	return nil
+}
+
+func WriteTypeDef(b io.StringWriter, f *winmd.Metadata, def *winmd.TypeDef) error {
+	switch def.Extends.Tag {
+	case coded.TypeDefOrRef_TypeRef:
+		// TODO: Keep track of this index rather than looking it up for each enum type.
+		record, err := f.Tables.TypeRef.Record(def.Extends.Index)
+		if err != nil {
+			return err
+		}
+		if record.Namespace.String() == "System" && record.Name.String() == "Enum" {
+			return writeTypeDefEnum(b, f, def)
+		}
+		return writeTypeDefStruct(b, f, def)
+	default:
+		return fmt.Errorf("unexpected type extends coded index: %#v", def.Extends)
+	}
+}
+
+func writeTypeDefEnum(b io.StringWriter, f *winmd.Metadata, def *winmd.TypeDef) error {
+	b.WriteString("type ")
+	b.WriteString(def.Name.String())
+
+	// Per §I.8.5.2 CLS Rule 7, the underlying type is the type of the field "__value". Find it.
+	var underlyingType *winmd.Type
+
+	type nameValuePair struct {
+		Name  string
+		Value any
+	}
+	// The number of enum members is the total number of fields minus the special "value__".
+	members := make([]nameValuePair, 0, def.FieldList.End-def.FieldList.Start-1)
+
+	maxNameLen := 0
+
+	for i := def.FieldList.Start; i < def.FieldList.End; i++ {
+		fd, err := f.Tables.Field.Record(i)
+		if err != nil {
+			return err
+		}
+		signature, err := f.FieldSignature(fd.Signature)
+		if err != nil {
+			return err
+		}
+		if fd.Name.String() == "value__" {
+			underlyingType = &signature.Type
+			continue
+		}
+
+		p := nameValuePair{fd.Name.String(), signature.Type}
+		members = append(members, p)
+		if len(p.Name) > maxNameLen {
+			maxNameLen = len(p.Name)
+		}
+	}
+	if underlyingType == nil {
+		return errors.New("failed to find underlying type for enum")
+	}
+
+	b.WriteString(" ")
+	if err := writeType(b, f, underlyingType); err != nil {
+		return err
+	}
+	b.WriteString("\n\nconst (\n")
+	for _, pair := range members {
+		b.WriteString("\t")
+		b.WriteString(pair.Name)
+		for i := 0; i < maxNameLen-len(pair.Name)+1; i++ {
+			b.WriteString(" ")
+		}
+		b.WriteString(def.Name.String())
+		b.WriteString(" = ")
+		// TODO: Look up enum member values in Constant table. Index values to avoid O(len(enums)*len(Constants))
+		b.WriteString("42")
+		b.WriteString("\n")
+	}
+	b.WriteString(")\n")
+	return nil
+}
+
+func writeTypeDefStruct(b io.StringWriter, f *winmd.Metadata, def *winmd.TypeDef) error {
+	b.WriteString("type ")
+	b.WriteString(def.Name.String())
+	b.WriteString(" struct {\n")
+
+	type nameTypePair struct {
+		Name string
+		Type *winmd.Type
+	}
+	fields := make([]nameTypePair, 0, def.FieldList.End-def.FieldList.Start)
+	maxNameLen := 0
+	for i := def.FieldList.Start; i < def.FieldList.End; i++ {
+		fd, err := f.Tables.Field.Record(i)
+		if err != nil {
+			return err
+		}
+		signature, err := f.FieldSignature(fd.Signature)
+		if err != nil {
+			return err
+		}
+		// Don't write anything yet, just save the necessary data. We need to wait until we know the
```

Comment:
The formatting business can be a rabbit hole, difficult to cover all corner cases, adds complexity and makes the code slower. IMO it would be easier for us to just write things that are formatted enough to latter be polished by `go/format`, as it is already done in [`cmd/genlayout#formatSource`.](https://github.com/microsoft/go-winmd/blob/fcfb7a086ae88d771d7dd23f7397630df451c8a1/cmd/genlayout/main.go#L56-L69)

Reply:
I noticed https://pkg.go.dev/go/format has a warning about this that I think applies in our situation: since it's a standard library but format can change, our results would then potentially change depending on the Go version that happened to be used to build our tool, which normally shouldn't happen. Could break end-to-end tests, or confuse people using this in `go generate` and getting more changes than they expect (potentially thrashing vs. other people 's changes if they also run the `go generate`). Doesn't seem likely to have a *big* impact for simple generated code like this, but something I noticed.

Agreed that post-formatting and simplifying the code is probably better than peppering in quirky formatting logic in various places.